# 1. Loading the data 

In [1]:
import pandas as pd 

In [ ]:
# Initially loading 1000 rows of each dataset to understand the structure and content
summary_df = pd.read_csv('job_summary.csv', nrows=1000)
skills_df = pd.read_csv('job_skills.csv', nrows=1000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=1000)


# 2. Exploring the data

## Looking at the first few rows 

In [5]:
print(f"First 5 rows of job_summary.csv:\n{summary_df.head()}\n\n\n")
print(f"First 5 rows of job_skills.csv:\n{skills_df.head()}\n\n\n")
print(f"First 5 rows of linkedin_job_postings.csv:\n{postings_df.head()}\n\n\n")


First 5 rows of job_summary.csv:
                                            job_link  \
0  https://www.linkedin.com/jobs/view/restaurant-...   
1  https://www.linkedin.com/jobs/view/med-surg-re...   
2  https://www.linkedin.com/jobs/view/registered-...   
3  https://uk.linkedin.com/jobs/view/commercial-a...   
4  https://www.linkedin.com/jobs/view/store-manag...   

                                         job_summary  
0  Rock N Roll Sushi is hiring a Restaurant Manag...  
1  Schedule\n: PRN is required minimum 12 hours p...  
2  Description\nIntroduction\nAre you looking for...  
3  Commercial account executive\nSheffield\nFull ...  
4  Address:\nUSA-CT-Newington-44 Fenn Road\nStore...  



First 5 rows of job_skills.csv:
                                            job_link  \
0  https://www.linkedin.com/jobs/view/housekeeper...   
1  https://www.linkedin.com/jobs/view/assistant-g...   
2  https://www.linkedin.com/jobs/view/school-base...   
3  https://www.linkedin.com/jobs/view/ele

## Count of dimensions

In [4]:
print(f"Shape of job_summary.csv: {summary_df.shape}")
print(f"Shape of job_skills.csv: {skills_df.shape}")
print(f"Shape of linkedin_job_postings.csv: {postings_df.shape}")

Shape of job_summary.csv: (1000, 2)
Shape of job_skills.csv: (1000, 2)
Shape of linkedin_job_postings.csv: (1000, 14)


## Column names

In [9]:
print(f"Column names in job_summay.csv: {summary_df.columns.tolist()}")
print(f"Column names in job_skills.csv: {skills_df.columns.tolist()}")
print(f"Column names in linkedin_job_postings.csv: {postings_df.columns.tolist()}")

Column names in job_summay.csv: ['job_link', 'job_summary']
Column names in job_skills.csv: ['job_link', 'job_skills']
Column names in linkedin_job_postings.csv: ['job_link', 'last_processed_time', 'got_summary', 'got_ner', 'is_being_worked', 'job_title', 'company', 'job_location', 'first_seen', 'search_city', 'search_country', 'search_position', 'job_level', 'job_type']


## Datatypes of each column

In [11]:
print(f"Job Summary columns datatypes:\n{summary_df.dtypes}\n{"-"*50}\n")
print(f"Job Skills columns datatypes:\n{skills_df.dtypes}\n{"-"*50}\n")
print(f"LinkedIn Job Postings columns datatypes:\n{postings_df.dtypes}\n{"-"*50}\n")

Job Summary columns datatypes:
job_link       str
job_summary    str
dtype: object
--------------------------------------------------

Job Skills columns datatypes:
job_link      str
job_skills    str
dtype: object
--------------------------------------------------

LinkedIn Job Postings columns datatypes:
job_link               str
last_processed_time    str
got_summary            str
got_ner                str
is_being_worked        str
job_title              str
company                str
job_location           str
first_seen             str
search_city            str
search_country         str
search_position        str
job_level              str
job_type               str
dtype: object
--------------------------------------------------



## Checking for missing values (First 1000 rows)

In [14]:
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link      0
job_skills    0
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                0
job_location           0
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 100000 rows)

In [15]:
summary_df = pd.read_csv('job_summary.csv', nrows=100000)
skills_df = pd.read_csv('job_skills.csv', nrows=100000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=100000)


In [18]:
print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link       0
job_skills    66
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                1
job_location           0
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 200000 rows)

In [20]:
summary_df = pd.read_csv('job_summary.csv', nrows=200000)
skills_df = pd.read_csv('job_skills.csv', nrows=200000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=200000)

print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")


Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link        0
job_skills    112
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                2
job_location           3
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 500000 rows)

In [3]:
summary_df = pd.read_csv('job_summary.csv', nrows=500000)
skills_df = pd.read_csv('job_skills.csv', nrows=500000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=500000)

print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link        0
job_skills    276
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link                0
last_processed_time     0
got_summary             0
got_ner                 0
is_being_worked         0
job_title               0
company                 3
job_location           10
first_seen              0
search_city             0
search_country          0
search_position         0
job_level               0
job_type                0
dtype: int64
 --------------------------------------------------



# 3. Cleaning the data

## Dropping columns with NULL skills

In [4]:
skills_df = skills_df.dropna(subset=['job_skills'])

print(f"Missing values in job skills dataset after dropping null job skills:\n {skills_df.isnull().sum()}")

Missing values in job skills dataset after dropping null job skills:
 job_link      0
job_skills    0
dtype: int64


## Fixing Iconsistent Text
- Lowercase all records
- Stripping whitespaces

In [5]:
# Job Titles
postings_df['job_title'] = postings_df['job_title'].str.lower().str.strip()
# Job Skills
skills_df['job_skills'] = skills_df['job_skills'].str.lower().str.strip()

# 4. Reducing/Filtering the data

## Numerosity Reduction

### Unique job titles and their counts

In [6]:
print(f"Unique job titles in linkedin job postings and their counts:\n {postings_df['job_title'].value_counts()}\n\n")

Unique job titles in linkedin job postings and their counts:
 job_title
customer service representative                                                                                                        3582
lead sales associate-ft                                                                                                                2963
store manager                                                                                                                          2523
assistant manager                                                                                                                      2195
first year tax professional                                                                                                            2115
                                                                                                                                       ... 
cub foods quarry - part time cake decorator - progressive scale from $15.19 to $17.50 / 

The dataset contains 250,825 unique job titles across all industries. However, for our project we are only interested in tech related job titles with sufficient postings to reliably analyze skill requirements.

### Filtering out all non-technical job titles

#### Creating list of technical job titles

In [13]:
tech_keywords = [
    # Software
    'software engineer', 'software developer', 'software architect', 'software designer', 
    
    # Web
    'frontend', 'front-end', 'backend', 'back-end', 'full stack', 'fullstack', 'web developer', 'web engineer', 'web designer', 
    
    # Data
    'data scientist', 'data analyst', 'data engineer', 'data architect', 'data visualization', 'data science', 'data analytics',
    
    # ML/AI
    'machine learning', 'deep learning', 'artificial intelligence', 'ai engineer', 'ml engineer', 'nlp', 'natural language processing', 'computer vision',
    
    # DevOps/Cloud
    'devops', 'cloud engineer', 'site reliability', 'sre', 'platform engineer', 'aws', 'azure', 'gcp', 'google cloud', 'amazon web services',
    
    # Mobile
    'android developer', 'ios developer', 'mobile developer', 'mobile engineer', 'mobile app developer', 'mobile application developer',
    
    # Security
    'cybersecurity', 'security engineer', 'penetration tester', 'security analyst', 'information security', 'infosec',
    
    # Database
    'database administrator', 'dba', 'database engineer', 'database developer', 'sql developer', 'sql engineer',
    
    # General
    'programmer', 'software qa', 'quality assurance engineer', 'test engineer', 'automation engineer', 'technical lead', 'tech lead', 'technology lead', 'cto', 'chief technology officer',
    
    # Common abbreviations safe to use
    'ui/ux', 'ux designer', 'ui designer',
    'it support', 'it specialist',
    'network engineer', 'blockchain'
]

#### Checking count of technical jobs in 500000 job postings

In [15]:
tech_count = postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False
).sum()

print(f"Tech related job postings: {tech_count}")
print(f"Out of total: {len(postings_df)}")

Tech related job postings: 26683
Out of total: 500000


We have 26683 technical jobs in total 500000 jobs 

#### Checking count of technical jobs in the entire dataset (1.3 Million job postings)

In [16]:
postings_df = pd.read_csv('linkedin_job_postings.csv')

In [18]:
tech_count = postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False
).sum()

print(f"Tech related job postings: {tech_count}")
print(f"Out of total: {len(postings_df)}")

Tech related job postings: 72385
Out of total: 1348454


So, in the total dataset, we have 72385 technical jobs in total 1.3 million job postings

#### Keeping only technical jobs and filtering out the rest

In [19]:
postings_df = postings_df[postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False)]

#### Shape after removing non-technical jobs

In [20]:
postings_df.shape

(72385, 14)

### Dimensionality Reduction

We are only interested in job_titles, hence, we remove all other attibutes

In [ ]:
postings_df = postings_df[['job_link' , 'job_title']]

print(f"Shape after filtering and selecting columns: {postings_df.shape}")